linalg_ops/iree-e2e-attention-test.cc

```diff
@@ -0,0 +1,491 @@
+      sum += exp(Attention[att_idx] - max_val);
+    }
+
+    // Apply softmax
+    for (int k2 = 0; k2 < K2; ++k2) {
+      int att_idx = index_3d(0, m, k2, M, K2);
+      Attention[att_idx] = exp(Attention[att_idx]) / sum;
+    }
+  }
+
+  // Compute Attention * V
+  for (int m = 0; m < M; ++m) {
+    for (int n = 0; n < N; ++n) {
+      float sum = 0.0;
+      for (int k2 = 0; k2 < K2; ++k2) {
+        int att_idx = index_3d(0, m, k2, M, K2);
+        int v_idx = index_3d(b, k2, n, K2, N);
+        sum += Attention[att_idx] * value_data[v_idx];
+      }
+      int o_idx = index_3d(b, m, n, M, N);
+      result_data[o_idx] = sum;
+    }
+  }
+}
+
+static iree_status_t reference_attention_element(
+    iree_hal_dim_t M, iree_hal_dim_t K1, iree_hal_dim_t K2, iree_hal_dim_t N,
+    iree_hal_dim_t B, iree_hal_element_type_t query_elem_type,
+    iree_hal_element_type_t key_elem_type,
+    iree_hal_element_type_t value_elem_type, void* query_data, void* key_data,
+    void* value_data, void* actual_data, void* result_data, iree_hal_dim_t b,
+    float* Attention) {
+  if (query_elem_type == IREE_HAL_ELEMENT_TYPE_FLOAT_32 &&
+      key_elem_type == IREE_HAL_ELEMENT_TYPE_FLOAT_32 &&
+      value_elem_type == IREE_HAL_ELEMENT_TYPE_FLOAT_32) {
+    reference_attention_f32_f32_f32_f32(
+        M, K1, K2, N, B, (const float*)query_data, (const float*)key_data,
+        (const float*)value_data, (float*)result_data, b, Attention);
+
+  } else {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "unhandled combination of element types in attention");
+  }
+  return iree_ok_status();
+}
+
+// Reference attention implementation, used to compare attention results
+// against.
+static iree_status_t reference_attention(
+    iree_hal_dim_t B, iree_hal_dim_t M, iree_hal_dim_t K1, iree_hal_dim_t K2,
+    iree_hal_dim_t N, iree_hal_element_type_t query_elem_type,
+    iree_hal_element_type_t key_elem_type,
+    iree_hal_element_type_t value_elem_type, iree_byte_span_t query_contents,
+    iree_byte_span_t key_contents, iree_byte_span_t value_contents,
+    iree_byte_span_t actual_contents, iree_byte_span_t result_contents,
+    int compute_every) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, B);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, M);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, K1);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, K2);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, N);
+
+  iree_host_size_t count = 0;
+  float* Attention = allocate_tensor(1, M, K2);
+  for (iree_hal_dim_t b = 0; b < B; ++b) {
+    if (++count < compute_every) continue;
+    count = 0;
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0,
+        reference_attention_element(
+            M, K1, K2, N, B, query_elem_type, key_elem_type, value_elem_type,
+            query_contents.data, key_contents.data, value_contents.data,
+            actual_contents.data, result_contents.data, b, Attention));
+  }
+  free_tensor(Attention);
```

Comment:
This leaks if an error occurs. Might be fine in a test, but could switch to a different code pattern:

```c++
iree_status_t status = iree_ok_status();

if (iree_status_is_ok(status)) {
  status = reference_attention_element(...);
}

free_tensor(...);
IREE_TRACE_ZONE_END(z0);
return status;
```

Or, since this is C++, you could use a smart pointer type instead of loose `malloc` and `free` calls.